projects/bootloader/scripts/flash_application_code.py

```diff
@@ -0,0 +1,85 @@
+# pylint: skip-file
+"""This script handles the client side of the flash application code protocol."""
+
+# Import protobuf
+from protogen import flash_application_code_pb2 as pb2
+from can_datagram import Datagram, DatagramListener, DatagramSender
+import can
+import time
+import os
+import sys
+currentdir = os.path.dirname(os.path.realpath(__file__))
+parentdir = os.path.dirname(currentdir)
+sys.path.append(parentdir)
+
+
+class FlashApplication():
+    """Flash Application Code Protocol"""
+
+    def __init__(self, board_ids, application_code_data: bytes, channel):
+        self.data = application_code_data
+        self.board_ids = board_ids
+        self.channel = channel
+
+        self.sender = DatagramSender(channel=self.channel, receive_own_messages=False)
+        self.listener = DatagramListener(self.listener_callback)
+        self.notifier = can.Notifier(self.sender.bus, [self.listener])
+
+    def listener_callback(self, datagram: Datagram, board_id):
+        """Check whether the datagram has an OK status code"""
+
+        if datagram._data != bytearray(0):
+            self.status_code = 1
+
+    def chunks(self, app_code: bytearray, chunk_size):
+        """Yield successive chunk_size byte-sized chunks from app_code."""
+        for i in range(0, len(app_code), chunk_size):
+            yield app_code[i:i + chunk_size]
+
+    def create_message(self, id, data):
+        message = Datagram(
+            datagram_type_id=id,
+            node_ids=self.board_ids,
+            data=bytearray(data))
+        return message
+
+    def flash_protobuf(self):
+        """Flash Protobuf"""
+
+        message = self.create_message(8, pb2.DESCRIPTOR.serialized_pb)
```

Comment:
Looks like you're sending a serialized description of the protobuf format itself here. Instead what you'll need to do is instantiate the protobuf and fill out its fields, which you can find in `flash_application_code.proto`:
```py
pb = pb2.FlashApplicationCode()
pb.name = ...
pb.git_version = ...
pb.application_crc = ...
pb.size = ...

# then to serialize:
serialized = pb.SerializeToString()
```
See https://developers.google.com/protocol-buffers/docs/pythontutorial#the-protocol-buffer-api. `name` and `git_version` will have to be passed into your class (that's the name of the project and the git commit hash it's from respectively). `size` is the size in bytes of the application code data, and `application_crc` is a hash that will have to be computed specially - talk to Faraz about how to compute it.